### PR TITLE
make cache file writing sync and hashed by query

### LINKED ISSUE #345: cache.json is large due to containing multiples copies of the same data
## Type of Change
- [ ] New Feature Request
- [ ] Documentation / Website
- [ ] Improvement / Suggestion
- [x] Bug
- [ ] Other (please clarify below)

## Summary
Moving the discussion from https://github.com/ProjectEvergreen/greenwood/issues/317#issuecomment-619227081, but after the `0.5.0` release, there are some serious concerns with _cache.json_ generation.  Docs page is transfering over 2.6MB of JSON data!

1. It's being over `fetch`ed (why!?)
1. It's huge!  Docs page
image

This also means all _index.html_ pages are huge too...
Screen Shot 2020-05-03 at 11 48 19 AM

## Details
Thoughts on quick wins here:
1. Fix over`fetch`ing by caching in the client. (but would also be good to know why this is happening)
1. Reduce duplication of query results within a single cache.json file (e.g. _about/cache.json_) because **deepmerge** just keeps doubling up the previous file contents, resulting in duplicates over and over again?

For something like _docs/cache.json_, this could bring the size down to 35k! We should also try and add a "budget" spec for this to make sure the size can't explode again without it failing at least.

> Later down the road, we should look to reduce duplication of query results across all _cache.json_ files - see #317

### PR DESCRIPTION
**Pull request by thescientist13**
## Related Issue
resolves #345 

## Summary of Changes
1. Write files by query, not by page
1. And write them synchronously to avoid overwriting

> For other mentioned issues #345, i will split them off into their own issues

### CODE REVIEW

**Comment by thescientist13** [comment]
Ugh... getting these stupid invariant errors in the Netlify preview environment. :/
Screen Shot 2020-05-06 at 8 29 42 PM

I don't know if it's related to this change though...?


Had a thought on it here https://github.com/ProjectEvergreen/greenwood/issues/331#issuecomment-624951466

**Comment by hutchgrant** [comment]
> Ugh... getting these stupid invariant errors in the Netlify preview environment. :/

Your Menu Query for the side menu is not being cached. Check cached files and you'll see theres only a navigation menu query, config, and graph queries.

**Comment by hutchgrant** [comment]
When using: 

```js
const md5 = crypto.createHash('md5').update(query).digest('hex');
```

I saw the following error:

"SyntaxError: /media/skynet/DATA/workspace/evergreen/greenwood/public/plugins/2bc8e256a25844b37c22af93673c67e3-cache.json: Unexpected token d in JSON at position 2830"

It's not being caught. I was only able to catch it by switching back to async fs.writeFile and fs.mkdirs.


If you switch back to:

```js
const md5 = crypto.createHash('md5').update(cache).digest('hex');
```

While still using synchronous as per this PR, it works fine, no errors.

**Comment by thescientist13** [comment]
> Your Menu Query for the side menu is not being cached. Check cached files and you'll see theres only a navigation menu query, config, and graph queries.

Oof...

> Notice how our testing didn't catch this? Because it relies on mock graph.

Yeah, noticed that too.  Will see if I can do something about that somehow.

> It's not being caught. I was only able to catch it by switching back to async fs.writeFile and fs.mkdirs.

Not quite sure I understand, but you are saying saw the JSON parse error when testing this branch and using `query`?  That's a bummer, I ran it a bunch of times on my local machine.  

So this might be a bit more complex of an issue, but I then since every _page_ gets serialized in parallel, all these GraphQL queries are happening as well.  I think it would be similar to what we did here for #271 ?

So the most sure fire way would be to make it all synchronous then.  Or i guess try the file locking approach?  Other options?

**Comment by hutchgrant** [comment]
The reason you aren't creating a side menu cache is simple: you're using the same data to generate a hash and therefore it only writes one file for two different queries!

menu query for side and navigation:
```js
query ($name: String, $route: String, $order: MenuOrderBy) {
  menu(name: $name, pathname: $route, orderBy: $order) {
    item {
      label
      link
      __typename
    }
    children {
      item {
        label
        link
        __typename
      }
      children {
        item {
          label
          link
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }
}

```


The fix is to include the query's variables to make it unique:
```js
const md5 = crypto.createHash('md5').update(JSON.stringify(query) + JSON.stringify(variables)).digest('hex');
```

File: www/components/shelf/shelf.js [hunk 1]
```diff
@@ -99,6 +83,32 @@ class Shelf extends LitElement {
...
+
+    this.shelfList = response.data.menu.children;
+    console.log('fetchShelfData - ', this.shelfList);
+  }
+
+  async updated(changedProperties) {
+    console.log('ENTER updated - changedProperties', changedProperties);
+    console.log('updated - this.page', this.page);
+    if (changedProperties.has('page') && this.page !== '' && this.page !== '/') {
+      await this.fetchShelfData();
```
Comment (hutchgrant) [inline www/components/shelf/shelf.js:105]: You're expecting to return a promise that doesn't exist.  Maybe just return the query(which is a promise) like this:

```js
const response = await this.fetchShelfData();
this.shelfList = response.data.menu.children;
```

File: www/templates/page-template.js [hunk 2]
```diff
@@ -26,12 +26,14 @@ class PageTemplate extends LitElement {
   }
 
   updated() {
```
Comment (hutchgrant) [inline www/templates/page-template.js:28]: My understanding is this lifecycle is run after the component has been rendered. I tested using the connectedCallback() to set the this.page variable from the page-template and had better results (I wasn't able to break it). Maybe that's just chance, I don't know.

File: www/templates/page-template.js [hunk 3]
```diff
@@ -26,12 +26,14 @@ class PageTemplate extends LitElement {
...
+    console.log('ENTER page template updated', window.location.pathname);
     this.route = window.location.pathname;
```
Comment (hutchgrant) [inline www/templates/page-template.js:30]: My understanding is this lifecycle is run after the component has been rendered. I tested using the connectedCallback() to set the this.page variable from the page-template and had better results (I wasn't able to break it). Maybe that's just chance, I don't know.

Comment (thescientist13) [reply www/templates/page-template.js:30]: `this.page` was actually undefined at that point, so `updated` seemed like a good way to make the sure the value is "ready".

File: www/components/shelf/shelf.js (hunk 1 above)
Comment (thescientist13) [reply www/components/shelf/shelf.js]: still happening, I guess I should just revert back the `connectedCallback` / `window.location.pathname` approach then? 

Not sure if we'll need to start recommending / documenting certain patterns around this stuff then, though I would hope we don't have to be any different than SPA development.  Like if having unpredictable lifecycles could be a concern, or just needing to guard initial values of your variables.

**Pull request merged**